### DISCUSSION TITLE
Type Generation: Types for Go

### DISCUSSION THREAD

**Discussion by isaacharrisholt** [supabase/supabase#20132] [Feature Requests]
I'd like to be able to generate types for Go in `postgres-meta`. This should be made easier by supabase/postgres-meta#685. My suggestion for a rough structure is as follows. Consider the following tables:

```sql
create table foo_table (
  id bigint not null primary key,
  name varchar(255)
);

create table bar_table (
  foo_id bigint not null references foo_table (id)
);

create table "another schema"."baz table" (
  -- etc.
);
```

The Go types would look something like:

```go
type PublicFooTable struct {
  Id   int64  `json:"id"`
  Name string `json:"name"`
}

type PublicBarTable struct {
  FooId int64 `json:"foo_id"`
}

type AnotherSchemaBazTable struct {
  // etc.
}
```

Essentially, the struct type would have the name `{schema}{tableName}` where both `schema` and `tableName` are both `PascalCase`. The same would apply to field names (they have to be capitalised at the very least as they can't be unmarshalled from JSON if they're not public).

This should work quite well with the `postgrest-go` package maintained by the community:

```go
var result struct {
	PublicBarTable
	Foo PublicFooTable `json:"foo"`
}

n, err := supabase.From("bar").
	Select("*, foo (*)").
        Single().
	ExecuteTo(&result)

// err check...

// Access fields:
barFooId := result.PublicBarTable.FooId
fooName := result.Foo.Name
```

It would also be great to be able to use table/column comments as [Go doc comments](https://tip.golang.org/doc/comment) so the resulting types can be nicely documented at both a struct and field level, e.g.:

```go
// user-defined comment for foo_table
type PublicFooTable struct {
  // user-defined comment for id
  Id   int64  `json:"id"`
  Name string `json:"name"`
}
```

This is not supported by the generator at the moment (I don't think, correct me if I'm wrong!) but it would be a great thing to implement.

Very open to suggestions here! If everyone's happy with the proposal, I'm happy to do the implementation work.

Thanks :)

**Comment by soedirgo** [comment]
Thanks for the proposal!

I have a feeling a generated client lib approach might work better for Go, but this is a good start and doesn't require much changes on postgrest-go 👍
